### PR TITLE
Support gang scheduling with Yunikorn

### PR DESCRIPTION
**Pull request by jacobsalway**
# Purpose

Add a scheduler implementation for [YuniKorn](https://yunikorn.apache.org) to support task group annotations on the driver pod for gang scheduling and queue labels on both driver and executor pods.

Completes the first dot point of https://github.com/kubeflow/spark-operator/issues/2098. Will add docs for the Spark operator in a separate PR, but will add docs to YuniKorn as well.

# Changes

* Add a new batch scheduler with the name `yunikorn`
* Calculate the resources required to schedule both driver and executor pods i.e. requests and executor counts, and use this to create task groups annotations on the driver pod to support [gang scheduling](https://yunikorn.apache.org/docs/next/user_guide/gang_scheduling/)
* Use the existing `batchSchedulerOptions.queue` to allow queue configuration. This is not a required field as Yunikorn supports [placement rules](https://yunikorn.apache.org/docs/0.8.0/configuration/placement_rules/) that allow an app to be placed without explicitly specifying a queue

# Testing

```bash
# clone my fork and checkout the branch
git clone git@github.com:jacobsalway/spark-operator.git
cd spark-operator && git checkout add-yunikorn-scheduler

# build the docker image
docker build . -t docker.io/kubeflow/spark-operator:add-yunikorn-scheduler

# create a kind cluster
kind create cluster --wait=1m

# load the operator image
kind load docker-image docker.io/kubeflow/spark-operator:add-yunikorn-scheduler

# install the operator with the forked build and the batch scheduler enabled
helm install spark-operator ./charts/spark-operator-chart \
  --create-namespace \
  --namespace spark-operator \
  --set image.tag=add-yunikorn-scheduler \
  --set controller.batchScheduler.enable=true

# install yunikorn
helm install yunikorn yunikorn \
  --repo https://apache.github.io/yunikorn-release \
  --create-namespace \
  --namespace yunikorn

# create the example application
kubectl apply -f examples/spark-pi-yunikorn.yaml

# observe that yunikorn will gang schedule the application
# task group pods prefixed with `tg-spark-...-spark-{driver/executor}-...` with requests
# that match the real pods will be created
```

https://github.com/user-attachments/assets/2e630b34-ef3c-457d-891f-5ad2bb438e12

## Change Category

Indicate the type of change by marking the applicable boxes:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- ~[ ] Documentation update~ will do in a separate PR to avoid this one getting any larger

## Checklist

Before submitting your PR, please review the following:

- [X] I have conducted a self-review of my own code.
- ~[ ] I have updated documentation accordingly.~ will do in a separate PR to avoid this one getting any larger
- [X] I have added tests that prove my changes are effective or that my feature works.
- [X] Existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by jacobsalway** [comment]
Still working on this, just sorting out some issues with converting the java byte string to Kubernetes resource values. Have re-read through the upstream code and have a better approach in mind

**Comment by jacobsalway** [comment]
Stacked PR also available to allow the default batch scheduler to be set if not specified by the user https://github.com/jacobsalway/spark-operator/pull/1

File: internal/scheduler/yunikorn/scheduler.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
+package yunikorn
```
Comment (ChenYi015) [inline internal/scheduler/yunikorn/scheduler.go:1]: We can add license comment in every new file, like https://github.com/kubeflow/spark-operator/blob/5972482ca4dbe059914a9b5b70edb98a82dfc0a6/internal/controller/sparkapplication/controller.go#L1-L15.

File: internal/scheduler/yunikorn/resourceusage/executor_count.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+		if app.Spec.DynamicAllocation.MinExecutors != nil {
+			initialExecutors = max(initialExecutors, *app.Spec.DynamicAllocation.MinExecutors)
+		}
+		if app.Spec.DynamicAllocation.InitialExecutors != nil {
+			initialExecutors = max(initialExecutors, *app.Spec.DynamicAllocation.InitialExecutors)
+		}
+	}
+
+	return initialExecutors
+}
```
[the remark this answers is not in the document]
Comment (ChenYi015) [reply internal/scheduler/yunikorn/resourceusage/executor_count.go:22]: Maybe we can move this func`NumInitialExecutors` to `pkg/util/sparkapplication.go` as this is an util function for SparkApplication.

File: internal/scheduler/yunikorn/scheduler.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+)
+
+const (
+	SchedulerName = "yunikorn"
+
+	// The names are set to match the Yunikorn gang scheduling example for Spark, but these can be any
+	// value as long as what's on the pod matches the task group definition
+	// https://yunikorn.apache.org/docs/next/user_guide/gang_scheduling/#enable-gang-scheduling-for-spark-jobs
+	driverTaskGroupName   = "spark-driver"
+	executorTaskGroupName = "spark-executor"
```
Comment (ChenYi015) [inline internal/scheduler/yunikorn/scheduler.go:21]: I see that in Volcano scheduler, the PodGroup names are different between SparkApplications. So I am wondering whether we need to set distinct task group names for different SparkApplications?

**Review by ChenYi015 — COMMENTED** [review]
Thanks for your contributions, I have left some comments.

File: internal/scheduler/yunikorn/scheduler.go (hunk 3 above)
Comment (jacobsalway) [reply internal/scheduler/yunikorn/scheduler.go:21]: For Yunikorn, a task group is a property of an application, so the task group names only need to be unique within the application they're a part of. 

Application IDs must be unique in the same way that a Volcano PodGroup name has to be, however there's a nice integration in Yunikorn to use the `spark-app-selector` label for the application ID.

File: internal/scheduler/yunikorn/scheduler.go (hunk 1 above)
Comment (jacobsalway) [reply internal/scheduler/yunikorn/scheduler.go:1]: I'll add the license comment to the top of all my new files, but what do you think if in a separate PR we add a license comment check as part of CI, and maybe a Makefile target for checking and adding locally?

File: internal/scheduler/yunikorn/resourceusage/executor_count.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jacobsalway) [reply internal/scheduler/yunikorn/resourceusage/executor_count.go:38]: Ah nice, good catch. I think I'd only been looking at https://github.com/apache/spark/blob/899fad4710bef174684deee64314ac483c16c494/core/src/main/scala/org/apache/spark/util/Utils.scala#L2534-L2557 when I wrote this implementation.

Will fix and move to utils tomorrow

[the remark this answers is not in the document]
Comment (jacobsalway) [reply internal/scheduler/yunikorn/resourceusage/executor_count.go]: Fixed to match the core implementation and moved to `pkg/utils/sparkapplication.go`

File: internal/scheduler/yunikorn/scheduler.go (hunk 1 above)
Comment (ChenYi015) [reply internal/scheduler/yunikorn/scheduler.go:17]: I think it is a good point to add license check as part of CI. cc @andreyvelich @vara-bonthu @yuchaoran2011 .

**Review by ChenYi015 — APPROVED** [review]

File: internal/scheduler/yunikorn/scheduler.go (hunk 1 above)
Comment (jacobsalway) [reply internal/scheduler/yunikorn/scheduler.go:17]: Created an issue for it https://github.com/kubeflow/spark-operator/issues/2139

**Review by yuchaoran2011 — APPROVED** [review]
/lgtm
Great to see that all the edge cases including min overhead memory are covered.

**Pull request merged**